src/System.IO/src/System/IO/BufferedStream.cs

```diff
@@ -0,0 +1,682 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Runtime.InteropServices;
+using System.Diagnostics.Contracts;
+
+namespace System.IO
+{
+    /// <summary>
+    /// One of the design goals here is to prevent the buffer from getting in the way and slowing
+    /// down underlying stream accesses when it is not needed. If you always read & write for sizes
+    /// greater than the internal buffer size, then this class may not even allocate the internal buffer.
+    /// See a large comment in Write for the details of the write buffer heuristic.
+    /// 
+    /// This class buffers reads & writes in a shared buffer.
+    /// (If you maintained two buffers separately, one operation would always trash the other buffer
+    /// anyways, so we might as well use one buffer.) 
+    /// The assumption here is you will almost always be doing a series of reads or writes, but rarely
+    /// alternate between the two of them on the same stream.
+    ///
+    /// Class Invariants:
+    /// The class has one buffer, shared for reading & writing.
```

Comment:
The three ampersands are being highlighted in red because they should be ```&amp;``` instead of ```&```.